concordium-smart-contract-testing/src/types.rs

```diff
@@ -0,0 +1,330 @@
+    pub balance:         AccountBalance,
+    /// Account policies.
+    pub policies:        v0::OwnedPolicyBytes,
+    /// The number of signatures. The number of signatures affect the cost of
+    /// every transaction for the account.
+    pub signature_count: u32,
+}
+
+/// An event that occurred during a contract update or invocation.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ChainEvent {
+    /// A contract was interrupted.
+    Interrupted {
+        /// The contract interrupted.
+        address: ContractAddress,
+        /// Logs produced prior to being interrupted.
+        logs:    v0::Logs,
+    },
+    /// A contract was resumed after being interrupted.
+    Resumed {
+        /// The contract resumed.
+        address: ContractAddress,
+        /// Whether the action that caused the interrupt succeeded.
+        success: bool,
+    },
+    /// A contract was upgraded.
+    Upgraded {
+        /// The contract upgraded.
+        address: ContractAddress,
+        /// The old module reference.
+        from:    ModuleReference,
+        /// The new module reference.
+        to:      ModuleReference,
+    },
+    /// A contract was updated.
+    Updated {
+        /// The contract updated.
+        address:    ContractAddress,
+        /// The name of the contract.
+        contract:   OwnedContractName,
+        /// The entrypoint called.
+        entrypoint: OwnedEntrypointName,
+        /// The amount added to the contract.
+        amount:     Amount,
+    },
+    /// A contract transferred an [`Amount`] to an account.
+    Transferred {
+        /// The sender contract.
+        from:   ContractAddress,
+        /// The [`Amount`] transferred.
+        amount: Amount,
+        /// The receiver account.
+        to:     AccountAddress,
+    },
+}
+
+/// A transfer from an contract to an account.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Transfer {
+    /// The sender contract.
+    pub from:   ContractAddress,
+    /// The amount transferred.
+    pub amount: Amount,
+    /// The receive account.
+    pub to:     AccountAddress,
+}
+
+/// Represents a successful deployment of a [`ContractModule`].
+#[derive(Debug, PartialEq, Eq)]
+pub struct SuccessfulModuleDeployment {
+    /// The reference of the module deployed.
+    pub module_reference: ModuleReference,
+    /// The energy used for deployment.
+    pub energy_used:      Energy,
+    /// Cost of transaction.
+    pub transaction_fee:  Amount,
+}
+
+/// An error that can occur while deploying a [`ContractModule`].
+// TODO: Can we get Eq for this when using io::Error?
+// TODO: Should this also have the energy used?
+#[derive(Debug, Error)]
+pub enum DeployModuleError {
+    /// Failed to read the module file.
+    #[error("could not read the file due to: {0}")]
+    ReadFileError(#[from] std::io::Error),
+    /// The module provided is not valid.
+    #[error("module is invalid due to: {0}")]
+    InvalidModule(#[from] anyhow::Error),
```

Comment:
Could we avoid exposing anyhow::Error?
Ideally it would be a proper defined error, but it would probably be a large task to introduce this, so that maybe something we should discuss